GH-1522: Made sure, diff manager works only on Git managed resources.

kittaakos opened the pull request:
Closes #1522.

Signed-off-by: Akos Kitta <kittaakos@gmail.com>

kittaakos commented:
The PR should be sufficient to avoid the `git show` related errors.

kittaakos commented on packages/git/src/node/dugite-git.spec.ts:
```diff
@@ -407,14 +414,14 @@ describe('git', async function () {
-                [4, 'six 🍏'],
-                [5, 'six 🍏'],
-            ]);
+                [
```
Sorry for messing up the apples... :(

kittaakos commented:
BTW, the actual fix was performed in [`dugite-extra`](https://github.com/theia-ide/dugite-extra/commit/e5263e2e486871186f560decc84584b82c0e886e).

AlexTugarev commented on packages/git/src/node/dugite-git.ts:
```diff
@@ -455,15 +455,22 @@ export class DugiteGit implements Git {
-        if (encoding === 'binary') {
-            return (await getBlobContents(repositoryPath, commitish, path)).toString();
+        const errorCodes = new Set([0, 1, 128]); // 128 to handle if the file is not in Git.
+        const result = await (async () => {
```
is it same as ...?
```
result = encoding === 'binary' 
 ? await getBlobContents() 
 : await getTextContents()
```

kittaakos replied:
Yes. I do not mind changing it.

AlexTugarev commented on packages/git/src/node/dugite-git.ts:
```diff
@@ -455,15 +455,22 @@ export class DugiteGit implements Git {
         const path = this.getFsPath(uri);
-        if (encoding === 'binary') {
-            return (await getBlobContents(repositoryPath, commitish, path)).toString();
+        const errorCodes = new Set([0, 1, 128]); // 128 to handle if the file is not in Git.
```
isn't 128 to general (in the code comment at least)?

kittaakos replied:
What would you expect? Just asking.

AlexTugarev replied:
I mean, I think it's right. but the comment might be misleading, as 128 is returned for all kind of errors, right?

AlexTugarev replied:
> What would you expect?

thirst thing I though reading this comment was, that this error case has a dedicated error code ;-) 
after looking at your changes to deguite-extra (https://github.com/theia-ide/dugite-extra/commit/e5263e2e486871186f560decc84584b82c0e886e#diff-ba3a11c6748e33b94d680245fe67866dR58), I understood that it's more like, if git crashes we return `undefined`.

kittaakos replied:
Kind of, there are a [few errors](https://github.com/desktop/dugite/blob/5c74239622fb896d3c3fa1a0f2d89ef0ce257981/lib/errors.ts#L2) that can be handled separately, but this is not among them.

AlexTugarev replied:
interesting hint! could we still parse the stdout provided by dugite to match that error?

e.g. 
```
> git show HEAD:unknown
fatal: Path 'unknown' does not exist in 'HEAD'
```

AlexTugarev replied:
that would be a more specific solution, but the more generic 128-code filter should work, too. I'm trying out.

kittaakos replied:
Sure, we can parse the error, but that implies, the error was already thrown. So with that approach, we would not suppress it. It worth a question from the dugite guys.

AlexTugarev replied:
do you mean that error here: https://github.com/theia-ide/dugite-extra/blob/e5263e2e486871186f560decc84584b82c0e886e/src/core/git.ts#L142 ?

AlexTugarev replied:
I see, I assumed it would be on that level: https://github.com/desktop/dugite/blob/master/examples/parsing-errors.ts#L9

AlexTugarev replied:
just to outline the option I've seen, after reading that error parsing example: instead of passing a set of acceptable codes, we could pass a second error code parser to `git` from `dugite-extra`, which would be chained after the existing one. that custom error parser could then test the `stderr`.

AlexTugarev replied:
but actually, there would be no more gain of that.

kittaakos replied:
I did a small experiment with different use-cases, it seems the error message can slightly vary based on the repository state:

```typescript
 describe('show', () => {
    it('exiting file', async () => {
      const testRepoPath = await initialize('desktop-show-existing')
      const filePath = Path.join(testRepoPath, 'file.txt')

      Fs.writeFileSync(filePath, 'some content', { encoding: 'utf8' })

      await GitProcess.exec(['add', '.'], testRepoPath)
      await GitProcess.exec(['commit', '-m', '"added a file"'], testRepoPath)

      const result = await GitProcess.exec(['show', ':file.txt'], testRepoPath)
      verify(result, r => {
        expect(r.exitCode).to.be.equal(0)
        expect(r.stdout.trim()).to.be.equal('some content')
      })
    })
    it('missing from index', async () => {
      const testRepoPath = await initialize('desktop-show-missing-index')

      const result = await GitProcess.exec(['show', ':missing.txt'], testRepoPath)
      verify(result, r => {
        expect(r.exitCode).to.be.equal(128)
        expect(
          r.stderr
            .trim()
            .endsWith(`Path 'missing.txt' does not exist (neither on disk nor in the index).`)
        ).to.be.true
      })
    })
    it('missing from HEAD', async () => {
      const testRepoPath = await initialize('desktop-show-missing-head')

      const filePath = Path.join(testRepoPath, 'file.txt')

      Fs.writeFileSync(filePath, 'some content', { encoding: 'utf8' })

      await GitProcess.exec(['add', '.'], testRepoPath)
      await GitProcess.exec(['commit', '-m', '"added a file"'], testRepoPath)

      const result = await GitProcess.exec(['show', 'HEAD:missing.txt'], testRepoPath)
      verify(result, r => {
        expect(r.exitCode).to.be.equal(128)
        expect(r.stderr.trim().endsWith(`Path 'missing.txt' does not exist in 'HEAD'`)).to.be.true
      })
    })
    it('missing from HEAD - no HEAD', async () => {
      const testRepoPath = await initialize('desktop-show-missing-head')

      const result = await GitProcess.exec(['show', 'HEAD:missing.txt'], testRepoPath)
      verify(result, r => {
        expect(r.exitCode).to.be.equal(128)
        expect(r.stderr.trim().endsWith(`Invalid object name 'HEAD'.`)).to.be.true
      })
    })
    it('-- missing', async () => {
      const testRepoPath = await initialize('desktop-show-missing-head')

      const filePath = Path.join(testRepoPath, 'file.txt')

      Fs.writeFileSync(filePath, 'some content', { encoding: 'utf8' })

      await GitProcess.exec(['add', '.'], testRepoPath)
      await GitProcess.exec(['commit', '-m', '"added a file"'], testRepoPath)

      const result = await GitProcess.exec(['show', '--', 'missing.txt'], testRepoPath)
      verify(result, r => {
        expect(r.exitCode).to.be.equal(0)
        expect(r.stdout.trim()).to.be.empty
      })
    })
    it('-- missing - outside', async () => {
      const testRepoPath = await initialize('desktop-show-missing-head')

      const filePath = Path.join(testRepoPath, 'file.txt')

      Fs.writeFileSync(filePath, 'some content', { encoding: 'utf8' })

      await GitProcess.exec(['add', '.'], testRepoPath)
      await GitProcess.exec(['commit', '-m', '"added a file"'], testRepoPath)

      const result = await GitProcess.exec(['show', '--', '/missing.txt'], testRepoPath)
      verify(result, r => {
        expect(r.exitCode).to.be.equal(128)
        expect(r.stderr.trim().endsWith('is outside repository')).to.be.true
      })
    })
  })
```

kittaakos replied:
I would do the followings:
 - Go with `128` as is now.
 - If the result is `undefined` we check the `stderr` **in the Theia code**, and if that contains information about the missing path, we ignore it. Otherwise, we can throw an error.

With this, we ignore the missing path in Theia, we keep the `dugite-extra` code as close to the GitHub Desktop code as possible. But we still throw an error if we suspect something unexpected error message from Git.

AlexTugarev commented on packages/git/src/browser/git-resource.ts:
```diff
@@ -17,10 +17,11 @@ export class GitResource implements Resource {
             const commitish = this.uri.query;
-            return await this.git.show(this.repository, this.uri.toString(), Object.assign({ commitish }, options));
+            const content = await this.git.show(this.repository, this.uri.toString(), Object.assign({ commitish }, options));
+            return content === undefined ? '' : content;
```
I've just tested that change, because I assume it has a implication on dirty diff manager (on master), but I was surprised by the value of `content`, it's `null`.

AlexTugarev replied:
Q: is it true, that `readContents` requires a result, or may it reject? 
if `readContents` may not reject, that implies a follow-up change, I kind of used that in dirty diff manager to detect that situation. if that `null` thing is sorted out, the previous revision state for dirty diff computation would be an empty string, witch would mark all contents of git-ingored files as added :-(

I think it's not a big deal, instead of using the git resource, we'd need to call `git.show` directly and check the result.

kittaakos replied:
What is your preference? If it can reject, we have to cover it all over the application. I would go with calling `git show` directly. What's your take on that?

AlexTugarev replied:
I searched for references of `Resource.readContents` and it looks like otherwise it's never considered it might resolve with `undefined`.

calling `git show` would fix that too. if you like, I'll create PR for yours.

kittaakos commented:
As discussed:
 - [ ] https://github.com/theia-ide/theia/pull/1562#discussion_r176144272
 - [ ] #1565?

kittaakos commented:
Given the fact that the error message could vary based on the file resource's path we would like to check with `git show` I would recommend to introduce a new method on the `git` API to check whether the file is under version control or not. So we would rely on Git, we do not have to parse error messages in `git show` and `git show` would not behave differently than other Git commands. What do you think @AlexTugarev? We need to catch the errors in the `GitResource` anyway.

From the diff manager's perspective, it would be something like this:
```typescript
if (!(await git.isVersionControld(repositoryPath, fileUri)) {
  // No need to handle it with the diff manager.
} else {
  // Do the same as we do now.
  git.show(repositoryPath, fileUri, ...);
}
```

AlexTugarev commented:
@kittaakos I'd like to have that explicit check! 

Q: should be got with `git.isVersionControlled`, or something like `git.isIgnored`? Not sure how that works, but I think forcibly added files should be treated as version controlled.

kittaakos commented:
I would like to keep the method names as close to the actual git commands as possible, we did that in the past too.

```typescript
    /**
     * Resolves to `true` if the file is managed by the Git repository. Otherwise, `false`.
     */
    lsFiles(repository: Repository, uri: string, options: { errorUnmatch: true }): Promise<boolean>;
    /**
     * Shows information about files in the index and the working tree
     *
     * @param repository the repository where the `git lf-files` has to be executed.
     * @param uri the URI of the file to check.
     * @param options further options for the command executions.
     */
    // tslint:disable-next-line:no-any
    lsFiles(repository: Repository, uri: string, options?: Git.Options.LsFiles): Promise<any>;
```

What do you think? Does it look too complicated? If so, I like your recommendation better than my original one.

AlexTugarev commented:
Looks good @kittaakos! I was assuming you'd go with `check-ignore`, but that's even better.

BTW it's `git ls-files`, right?

kittaakos commented:
No, not with `check-ignore`. An untarcked file is not ignored, still is not version controlled, so you cannot call `git show` on it. I would go with `git ls-files --errorUnmatch`. Do you have a better approach?

kittaakos commented:
@AlexTugarev, I have rewritten this and I've also reverted the [`dugite-extra` changes](https://github.com/theia-ide/dugite-extra/commit/ab0143311b06a30837166b2c11b5b71db63f0e28). Please review. Thanks!

AlexTugarev commented on packages/git/src/browser/dirty-diff/dirty-diff-manager.ts:
```diff
@@ -136,6 +139,18 @@ export class DirtyDiffManager {
+            return false;
+        }
+        const errorUnmatch = true;
+        return this.git.lsFiles(repository, uri, { errorUnmatch });
```
I think that will break some cases. I'd rather push that check down to the models, but let them create for the opened editor: 
https://github.com/theia-ide/theia/blob/9594d361045edb253232177d6d3e60e596f6dc3b/packages/git/src/browser/dirty-diff/dirty-diff-manager.ts#L228-L232. 

A file which is not version controlled can be opened in the editor, then you can `git add` it to stage, and then modify it. Current implementation creates models for all opened `file` editors. FYI, in #1526, there will be more improvements regarding reading of resources on demand.

kittaakos replied:
That is a good idea. Do you have any objections changing the method to `async`?

AlexTugarev replied:
All right.

AlexTugarev commented on packages/git/src/browser/dirty-diff/dirty-diff-manager.ts:
```diff
@@ -136,6 +139,18 @@ export class DirtyDiffManager {
+        if (new URI(uri).scheme !== 'file') {
+            return false;
+        }
+        const repository = this.repositoryTracker.selectedRepository;
```
I'm not sure how achieve this, but could we open a file in the editor, and the `selectedRepository` will be first `undefined`, but if this changes afterwards, I'd like to see the dirty diffs.

kittaakos replied:
I can understand your concern about this "enhancement", but that is unrelated to the `fatal ... is outside repository` issue this PR is intended to solve. Or did I miss something?

AlexTugarev replied:
Yes, I tried to make it clearer in the other comment. The manager should not exclude file URIs if the repository is not available, or the file is not version controlled, as it might change later on, right?

kittaakos commented:
I have updated the PR based on your feedback. Please review, thank you! (It is not urgent at all.)

AlexTugarev commented on packages/git/src/browser/dirty-diff/dirty-diff-manager.ts:
```diff
@@ -202,18 +205,19 @@ export class DirtyDiffModel implements Disposable {
             this.previousContent = [];
         }
-        if (noRelevantChanges && this.isInGitRepository(repository)) {
+        const inGitRepository = await this.isInGitRepository(repository);
```
just a minor, but could we push the "expensive" one level deeper after evaluation of `noRelevantChanges`?
